### PR TITLE
quic: make server connection defer sending packets during processing incoming packets

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: enable defer sending in QuicConnection::ProcessUdpPacket() on server side. This bundles packet writes for each connection at the end of the event loop instead of each packet read. It is for improving CPU cost of sendmsg() for QUIC listener, especially with UDP GSO batch writing.

Also change how closing connections are deferred during L4 filters creation. Without this change, the connection close is deferred till handling all types of delayed close in `QuicSession::OnCanWrite()` which `ProcessUdpPacket()` would call into upon handling the 1st incoming packet. With this change, connection close caused by L4 filter creation is handled right before calling ProcessUdpPacket() because OnCanWrite() is deferred.

Risk Level: medium, behavior change of quic write pipeline  
Testing: existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Runtime guard: envoy_reloadable_features_quic_defer_send_in_response_to_packet

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
Looks like //test/integration:quic_protocol_integration_test is failing. Is that expected?

File: source/common/quic/envoy_quic_server_connection.cc
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_server_connection.cc:24]: out of curiosity, instead of adding a new envoy-specific member, would it make sense to expose `defer_send_in_response_to_packets_` via a getter on QuicConnection?

File: source/common/quic/envoy_quic_server_session.cc
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_server_session.cc:194]: Out of curiosity, would we want the Client session to have the same logic that the server session adds in this PR?

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (RyanTheOptimist) [inline source/common/quic/quic_filter_manager_connection_impl.cc:152]: Do you think there's any need for a feature flag to control this part of the change? Presumably you're confident that it's a safe change? Or should we guard it with the same flag?

**Review by RyanTheOptimist — COMMENTED** [review]
I might be missing it, but do we have test which verifies the intended behavior for this change? Specifically that this "bundles packet writes for each connection at the end of the event loop instead of each packet read. " I was expecting to see a test were multiple packets were read and the flag controlled when packets were sent?

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (danzh2010) [reply source/common/quic/quic_filter_manager_connection_impl.cc:152]: It's guarded by close_type_during_initialize_. If the current feature flag is on, close_type_during_initialize_ wont' have value here, as OnCanWrite() happens outside of ProcessUdpPacket().

File: source/common/quic/envoy_quic_server_session.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_session.cc:194]: We are not doing this in any existing QUICHE client. Maybe client CPU usage hasn't been a problem.

File: source/common/quic/envoy_quic_server_connection.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_connection.cc:24]: That's also possible.

**Comment by danzh2010** [comment]
> I might be missing it, but do we have test which verifies the intended behavior for this change? Specifically that this "bundles packet writes for each connection at the end of the event loop instead of each packet read. " I was expecting to see a test were multiple packets were read and the flag controlled when packets were sent?

I modified test/common/quic/active_quic_listener_test.cc to check that send alarm is armed after receiving CHLO.

**Comment by danzh2010** [comment]
> Looks like //test/integration:quic_protocol_integration_test is failing. Is that expected?

Nope, I'm looking into it. They are all upstream disconnect tests, but it's weird that they fail only in windows build.

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (RyanTheOptimist) [reply source/common/quic/quic_filter_manager_connection_impl.cc:152]: I don't think I understand. `maybeApplyDelayClosePolicy()` is called from a couple of different places. Before this PR, it's possible for `maybeApplyDelayClosePolicy()` to actually call `close()` if not in delayed close (on line 150). But with this PR, if we're not in delayed close then `maybeApplyDelayClosePolicy()` will not call `close()`.

But I guess you're saying that before this PR, `close_type_during_initialize_` will never have a value here. Do I have that right?

File: source/common/quic/envoy_quic_server_connection.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_server_connection.cc:24]: No need to block this PR, but care to cook up a quiche CL and eventually switch over to using it?

**Comment by RyanTheOptimist** [comment]
> > I might be missing it, but do we have test which verifies the intended behavior for this change? Specifically that this "bundles packet writes for each connection at the end of the event loop instead of each packet read. " I was expecting to see a test were multiple packets were read and the flag controlled when packets were sent?
> 
> I modified test/common/quic/active_quic_listener_test.cc to check that send alarm is armed after receiving CHLO.

*nod* And presumably the send alarm being set leads to the behavior mentioned in the PR description. But it's not obvious to me that this is the case. (Though I certainly believe you that it is!) Would it be hard to have a test which actually waited to send something until multiple packets are received? For example, the existing test in active_quic_listener_test.cc doesn't seem to send any packets after receiving the CHLO. Is there a way to make it do that?

File: source/common/quic/envoy_quic_server_connection.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_server_connection.cc:24]: Oh, you already landed this CL! Nice!

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (danzh2010) [reply source/common/quic/quic_filter_manager_connection_impl.cc:152]: maybeApplyDelayClosePolicy() is only called in OnCanWrite() which was called at the end of ProcessUdpPacket() and various other places before this PR. But in this PR, we check `close_type_during_initialize_` at the beginning of ProcessUdpPacket() and close connection there. So OnCanWrite() won't be called at all in ProcessUdpPacket(). This check only takes effect at processing the 1st packet, `close_type_during_initialize_` will always be nullptr afterwards.

Comment (RyanTheOptimist) [reply source/common/quic/quic_filter_manager_connection_impl.cc:152]: Ok, I see. Thanks for the explanation.

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (RyanTheOptimist) [inline source/common/quic/quic_filter_manager_connection_impl.cc:102]: Presumably you're still debugging the CI failures?

Comment (danzh2010) [reply source/common/quic/quic_filter_manager_connection_impl.cc:102]: Yes, this is the only way I could think of to debug windows.

**Comment by danzh2010** [comment]
> _nod_ And presumably the send alarm being set leads to the behavior mentioned in the PR description. But it's not obvious to me that this is the case. (Though I certainly believe you that it is!) Would it be hard to have a test which actually waited to send something until multiple packets are received? For example, the existing test in active_quic_listener_test.cc doesn't seem to send any packets after receiving the CHLO. Is there a way to make it do that?

It was harder than I expected. CHLO doesn't always elicit sending packet due to packet coalescing, so there might be no behavioral change in this test with and without defer sending. But we are not testing QUICHE code, so to make the test more obvious, I used the getter of defer_send_in_response_to_packets_ to check its value in the test.

File: source/common/quic/envoy_quic_server_connection.cc
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_server_connection.cc:27]: nit: Mind filing a github issue to track this and adding a TODO to get to the bottom of this?

**Review by RyanTheOptimist — DISMISSED** [review]
LGTM modulo a TODO

File: source/common/quic/envoy_quic_server_connection.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_connection.cc:27]: done

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**
